### PR TITLE
HTCONDOR-1494-advertise-docker-cache

### PR DESCRIPTION
**Pull request by GregThain**
[HTCONDOR-1494](https://opensciencegrid.atlassian.net/browse/HTCONDOR-1494)

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [x] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [x] Verify that the branch destination of the PR matches the target version of the ticket
- [x] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [x] Check for documentation, if needed
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

[HTCONDOR-1494]: https://opensciencegrid.atlassian.net/browse/HTCONDOR-1494?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/condor_startd.V6/ResAttributes.cpp [hunk 1]
```diff
@@ -521,6 +527,20 @@ MachAttributes::compute_for_policy()
 	}
 }
 
+void 
+MachAttributes::compute_docker_cache() {
+#ifdef LINUX
+	static time_t lastRun = 0;
+	time_t now = time(nullptr);
+	if ((now - lastRun) > 1200) {
```
Comment (tannenba) [inline src/condor_startd.V6/ResAttributes.cpp:535]: Perhaps this '1200' should be a config knob, and/or at the very least the documentation should state that the docker cache size attribute is updated every 20 minutes

File: src/condor_utils/docker-api.cpp [hunk 2]
```diff
@@ -1466,6 +1468,149 @@ makeHostname(ClassAd *machineAd, ClassAd *jobAd) {
...
+	}
+
+	// sort the array by image name
+	std::sort(image_sizes.begin(), image_sizes.end());
+	
+	std::string imageFilename;
+
+	if( ! param( imageFilename, "LOG" ) ) {
+		dprintf(D_ALWAYS, "LOG not defined in param table, giving up\n");
+		return 0;
```
Comment (tannenba) [inline src/condor_utils/docker-api.cpp:1566]: Should not return 0 here, should return some negative number.  Also would be nice if the dprintf stated the function name DockerAPI:imageCacheUsed so we know what is "giving up".

File: src/condor_utils/docker-api.cpp [hunk 3]
```diff
@@ -1466,6 +1468,149 @@ makeHostname(ClassAd *machineAd, ClassAd *jobAd) {
...
+	}
+
+	imageFilename += "/.startd_docker_images";
+
+	std::vector<std::pair<std::string, int64_t>> images_on_disk;
+
+	FILE *f = safe_fopen_wrapper_follow(imageFilename.c_str(), "r");
```
Comment (tannenba) [inline src/condor_utils/docker-api.cpp:1573]: Don't you need a READ_LOCK on /.startd_docker_images ?  In gc_image() we get a WRITE_LOCK on this file...

**Review by tannenba — CHANGES_REQUESTED** [review]
Please address the 3 pending comments, thanks.

File: src/condor_startd.V6/ResAttributes.cpp (hunk 1 above)
Comment (GregThain) [reply src/condor_startd.V6/ResAttributes.cpp]: Knobbed and documented.

File: src/condor_utils/docker-api.cpp (hunk 3 above)
Comment (GregThain) [reply src/condor_utils/docker-api.cpp:1583]: read lock added.

**Review by tannenba — APPROVED** [review]
Thank you for the changes after the initial code review, LGTM

**Pull request merged**